### PR TITLE
Arrow schema converter.

### PR DESCRIPTION
**Pull request by liurenjie1024**
This is the first step of adding an arrow reader and writer for parquet-rs. 
This commit contains a converter which converts parquet schema to arrow schema.

### CODE REVIEW

**Comment by sunchao** [comment]
Thanks @liurenjie1024 . Will take a look soon. 

Meanwhile, I created #186 to track the overall progress of adding Arrow support. Please feel free to create more tasks in it. I'm also planning to take up some tasks. :)

**Review by sunchao — COMMENTED** [review]
Thanks @liurenjie1024 . Left some comments.

File: src/arrow_format/schema.rs
Comment (liurenjie1024) [reply src/arrow_format/schema.rs]: - I think panic when it's primitive type is same.
- I don't think we should panic when the `fields.len() != 1` because this is an error format and should not cause our function to panic.

File: src/arrow_format/schema.rs
Comment (liurenjie1024) [reply src/arrow_format/schema.rs]: They are not the same. The list item name must be equal to `{schema.name()}_tuple`

File: src/arrow_format/schema.rs
Comment (liurenjie1024) [reply src/arrow_format/schema.rs]: So that the user can just pass a vec![1,2,3] without calling .iter(). BTW, if T implements Iterator, it also implements IntoIterator.

**Comment by liurenjie1024** [comment]
@sunchao Thanks for the review and I've fixed the comments. As for the format part, how do you think about a rustfmt file so that other contributors can follow?

File: src/arrow_format/schema.rs
Comment (sunchao) [reply src/arrow_format/schema.rs]: You're right. For the latter maybe we can throw "not yet supported" error.

File: src/arrow_format/schema.rs
Comment (sunchao) [reply src/arrow_format/schema.rs]: I see. Looks good.

Comment (liurenjie1024) [reply src/arrow_format/schema.rs]: Yes, I didn't notice the check you mentioned before. I think we need another refactor so that we can share the code.

File: src/arrow_format/schema.rs
Comment (sadikovi) [reply src/arrow_format/schema.rs]: Why can't we return Err here?

Comment (liurenjie1024) [reply src/arrow_format/schema.rs]: Because if this happens, it means our code has bug.

Comment (sadikovi) [reply src/arrow_format/schema.rs]: To start with, returning an Err does not necessarily mean the code has bug. Err is just a way of expressing an invalid state or a bug and returning it without panicking. With Err it is easier to add a unit test to show that we return a proper error message.

By the way, there is no test for this, either with panic or Err. Also, what does it mean "It should not happen". What should not happen? Can you update the error message?

Comment (sadikovi) [reply src/arrow_format/schema.rs]: If you want to go with panic, I would suggest including something like the actual schema that was compared, and some other metadata as part of the error message, so it is easier to debug later on when it happens.

File: src/errors.rs
Comment (sadikovi) [inline src/errors.rs]: Can you add new line here?

File: src/arrow/schema.rs
Comment (sadikovi) [inline src/arrow/schema.rs]: Why do you need this function? How could self be part of self.leaves?

File: src/arrow/schema.rs
Comment (sadikovi) [inline src/arrow/schema.rs]: `) -> Result<Schema>` normally goes onto new line.

File: src/arrow/mod.rs
Comment (sadikovi) [inline src/arrow/mod.rs]: Empty lines.

File: Cargo.toml
Comment (sadikovi) [inline Cargo.toml]: Spaces around `=` - you have an example just above this line!

File: src/arrow/mod.rs
Comment (sadikovi) [inline src/arrow/mod.rs]: I would probably add a doc test here, showing how to use the function. Optional, but would be nice to have.

File: src/arrow/schema.rs
Comment (sadikovi) [inline src/arrow/schema.rs]: Why is `leaves` an `Rc<HashSet>`?

File: src/arrow/schema.rs
Comment (sadikovi) [inline src/arrow/schema.rs]: Is repeated field nullable? Normally it is complemented by nested optional or required fields. I am not very familiar with arrow, maybe it treats them as nullable by default.

File: src/arrow/schema.rs
Comment (sadikovi) [inline src/arrow/schema.rs]: `basic_info.repetition() == Repetition::REPEATED`.

File: src/arrow/schema.rs
Comment (sadikovi) [inline src/arrow/schema.rs]: Please, extend these error messages, they should indicate the error and potentially include metadata that could aid debugging.

File: src/arrow/schema.rs
Comment (sadikovi) [inline src/arrow/schema.rs]: Could you also check test utils? We might already have a function there to generate schemas.

File: src/arrow/schema.rs
Comment (sadikovi) [inline src/arrow/schema.rs]: We normally format it like this:
```rust
func(
  arg1,
  arg2,
  ...
)
```

File: src/schema/types.rs
Comment (sadikovi) [inline src/schema/types.rs]: I think it would be good to move the assertion into `column_root_of` method and return either Rc directly or ref extracted from Rc. In this case we do not need to duplicate the error message.

**Review by sadikovi — COMMENTED** [review]
Thanks for the PR. I left a few comments. Could you also add doc for the functions even if they are trivial, it is unclear what some of them do and why they do it, and will be difficult to maintain the code later.

File: src/arrow/schema.rs
Comment (liurenjie1024) [reply src/arrow/schema.rs]: This function is used to test if the schema included in the leaves. We need it when the converter is converting a primitive schema. The leaves are not leaves of the schema, but columns that need to convert.

File: src/arrow/schema.rs
Comment (liurenjie1024) [reply src/arrow/schema.rs]: I think the name `leaves` is quite misleading. It's not the leaves of schema, but the leave columns that needs to convert. I'll change that.

File: src/arrow/schema.rs
Comment (liurenjie1024) [reply src/arrow/schema.rs]: Yes, you're right. In fact there's no clear definition about the mapping to arrow and I just think this is default case.

File: src/arrow/schema.rs
Comment (liurenjie1024) [reply src/arrow/schema.rs]: If the repetition is not defined, this may cause panic.

Comment (sadikovi) [reply src/arrow/schema.rs]: I meant white spaces around it. I know that repetition unwraps the option.

File: src/arrow/schema.rs
Comment (liurenjie1024) [reply src/arrow/schema.rs]: Thanks. I found a parse which can convert texts to type. I'll change tests to use that.

**Comment by liurenjie1024** [comment]
@sadikovi Sorry for late reply. I've fixed some comments.

**Comment by liurenjie1024** [comment]
@sunchao @sadikovi Could you help to review this?

**Comment by sunchao** [comment]
@liurenjie1024 sure, will take a look soon. Sorry for the delay.

File: src/arrow/mod.rs
Comment (sadikovi) [reply src/arrow/mod.rs]: Okay, but let's not forget to do that.

**Review by sadikovi — COMMENTED** [review]
Looks good.

**Review by sunchao — APPROVED** [review]
Sorry for the late review (was on vacation). LGTM.

**Comment by liurenjie1024** [comment]
@sunchao @sadikovi We have two LGTM, help to merge this?

**Comment by liurenjie1024** [comment]
Yes, I think it would be better to merge it after merging with arrow.

**Comment by sunchao** [comment]
Great. We can create an umbrella JIRA after the merge, to track the parquet-arrow integration. I can help to get this PR in very soon after that.

**Comment by liurenjie1024** [comment]
Cool.

**Comment by sunchao** [comment]
@liurenjie1024 I created a JIRA here: https://issues.apache.org/jira/browse/ARROW-4060. Could you file a PR in the arrow repo for this? Thanks.

**Comment by sunchao** [comment]
Closing this as ARROW-4060 is already merged into Arrow.

**Pull request closed without merging**
